backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/extraction/impl/HibernateSearchMultiCollectorManager.java

```diff
@@ -0,0 +1,233 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.lucene.search.extraction.impl;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.hibernate.search.backend.lucene.lowlevel.collector.impl.CollectorExecutionContext;
+import org.hibernate.search.backend.lucene.lowlevel.collector.impl.CollectorFactory;
+import org.hibernate.search.backend.lucene.lowlevel.collector.impl.CollectorKey;
+import org.hibernate.search.backend.lucene.search.timeout.impl.LuceneCounterAdapter;
+import org.hibernate.search.engine.common.timing.Deadline;
+import org.hibernate.search.engine.search.timeout.spi.TimeoutManager;
+import org.hibernate.search.util.common.AssertionFailure;
+
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.CollectorManager;
+import org.apache.lucene.search.MultiCollector;
+import org.apache.lucene.search.TimeLimitingCollector;
+import org.apache.lucene.util.Counter;
+
+public abstract class HibernateSearchMultiCollectorManager
+		implements CollectorManager<Collector, HibernateSearchMultiCollectorManager.MultiCollectedResults> {
+
+	protected HibernateSearchMultiCollectorManager(TimeoutManager timeoutManager) {
+		this.timeoutManager = timeoutManager;
+	}
+
+	public static HibernateSearchMultiCollectorManager create(
+			TimeoutManager timeoutManager, Map<CollectorKey<?, ?>, CollectorManager<Collector, ?>> collectorManagers) {
+		if ( collectorManagers.size() == 1 ) {
```

Comment:
Might as well keep the comment, to understand why we're making code more complicated :x 

```suggestion
		if ( collectorManagers.size() == 1 ) {
			// avoid using a multi collector if we have just one collector,
			// as MultiCollector explicitly ignores the total hit count optimization
```